File: packages/flet/lib/src/services/tester.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';

import '../utils/keys.dart';

class TesterService extends FletService {
  TesterService({required super.control});
  final Map<int, TestFinder> _finders = {};

  @override
  void init() {
    super.init();
    debugPrint("Tester(${control.id}).init: ${control.properties}");
    control.addInvokeMethodListener(_invokeMethod);
  }

  @override
  void dispose() {
    debugPrint("Tester(${control.id}).dispose()");
    control.removeInvokeMethodListener(_invokeMethod);
    super.dispose();
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    debugPrint("Tester.$name($args)");
    switch (name) {
      case "pump":
        await control.backend.tester!
            .pump(duration: parseDuration(args["duration"]));

      case "pump_and_settle":
        await control.backend.tester!
            .pumpAndSettle(duration: parseDuration(args["duration"]));

      case "find_by_text":
        var finder = control.backend.tester!.findByText(args["text"]);
        _finders[finder.id] = finder;
        return finder.toMap();

      case "find_by_text_containing":
        var finder =
            control.backend.tester!.findByTextContaining(args["pattern"]);
        _finders[finder.id] = finder;
        return finder.toMap();

      case "find_by_key":
        var controlKey = parseKey(args["key"])!;
        var key = controlKey is ControlScrollKey
            ? control.backend.globalKeys[controlKey.toString()]
            : ValueKey(controlKey.value);
        if (key == null) {
          throw Exception("Key not found: $key");
        }
        var finder = control.backend.tester!.findByKey(key);
        _finders[finder.id] = finder;
        return finder.toMap();

      case "find_by_tooltip":
        var finder = control.backend.tester!.findByTooltip(args["value"]);
        _finders[finder.id] = finder;
        return finder.toMap();

      case "find_by_icon":
        var iconCode = args["icon"];
        var icon = parseIconData(iconCode, control.backend);
        if (icon == null) {
          throw Exception("Icon not found: $iconCode");
        }
        var finder = control.backend.tester!.findByIcon(icon);
        _finders[finder.id] = finder;
        return finder.toMap();

      case "take_screenshot":
        return await control.backend.tester!.takeScreenshot(args["name"]);

      case "tap":
        var finder = _finders[args["finder_id"]];
        if (finder != null) {
          await control.backend.tester!.tap(finder, args["finder_index"]);
        }

      case "long_press":
        var finder = _finders[args["finder_id"]];
        if (finder != null) {
          await control.backend.tester!.longPress(finder, args["finder_index"]);
        }

      case "enter_text":
        var finder = _finders[args["finder_id"]];
        if (finder != null) {
          await control.backend.tester!
              .enterText(finder, args["finder_index"], args["text"]);
        }

      case "mouse_hover":
        var finder = _finders[args["finder_id"]];
        if (finder != null) {
          await control.backend.tester!
              .mouseHover(finder, args["finder_index"]);
        }

      case "teardown":
        control.backend.tester?.teardown();

      default:
        throw Exception("Unknown Tester method: $name");
    }
  }
}

File: sdk/python/packages/flet/src/flet/controls/services/tester.py
from typing import Optional

from flet.controls.base_control import control
from flet.controls.duration import Duration
from flet.controls.keys import KeyValue
from flet.controls.services.service import Service
from flet.controls.types import IconValue
from flet.testing.finder import Finder

__all__ = ["Tester"]


@control("Tester")
class Tester(Service):
    async def pump(self, duration: Optional[Duration] = None):
        return await self._invoke_method_async("pump", {"duration": duration})

    async def pump_and_settle(self):
        return await self._invoke_method_async("pump_and_settle")

    async def find_by_text(self, text: str):
        finder = await self._invoke_method_async("find_by_text", {"text": text})
        return Finder(**finder)

    async def find_by_text_containing(self, text: str):
        finder = await self._invoke_method_async(
            "find_by_text_containing", {"text": text}
        )
        return Finder(**finder)

    async def find_by_key(self, key: KeyValue):
        finder = await self._invoke_method_async("find_by_key", {"key": key})
        return Finder(**finder)

    async def find_by_tooltip(self, value: str):
        finder = await self._invoke_method_async("find_by_tooltip", {"value": value})
        return Finder(**finder)

    async def find_by_icon(self, icon: IconValue):
        finder = await self._invoke_method_async("find_by_icon", {"icon": icon})
        return Finder(**finder)

    async def tap(self, finder: Finder):
        await self._invoke_method_async("tap", {"id": finder.id})

    async def enter_text(self, finder: Finder, text: str):
        await self._invoke_method_async("enter_text", {"id": finder.id, "text": text})

    def teardown(self):
        return self._invoke_method_async("teardown")
